app/src/main/java/org/mozilla/fenix/home/blocklist/BlocklistMiddleware.kt

```diff
@@ -0,0 +1,170 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.home.blocklist
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.map
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import mozilla.components.lib.state.Store
+import org.mozilla.fenix.home.HomeFragmentAction
+import org.mozilla.fenix.home.HomeFragmentState
+import org.mozilla.fenix.home.recentbookmarks.RecentBookmark
+import org.mozilla.fenix.home.recenttabs.RecentTab
+import org.mozilla.fenix.home.recentvisits.RecentlyVisitedItem
+
+class BlocklistMiddleware(
+    coroutineScope: CoroutineScope,
+    private val provider: HomescreenBlocklistProvider
+) : Middleware<HomeFragmentState, HomeFragmentAction> {
+
+    private var blocklist: Set<String> = setOf()
+    private var blocklistLoaded = false
+    private var store: Store<HomeFragmentState, HomeFragmentAction>? = null
+
+    init {
+        provider.blocklist.map { updatedList ->
+            blocklist = updatedList
+            if (!blocklistLoaded) {
+                filterLatestState()
+            }
+            blocklistLoaded = true
+        }.launchIn(coroutineScope)
```

Comment:
This might be a style thing, but I think we typically do `scope.launch { }` or some variation